### ISSUE TITLE
sql-hint.js doesn't work with multiplex.js

### ISSUE THREAD

**Issue by neon-dev** [codemirror/codemirror5#5249]
SQL autocompletion breaks when used in a multiplex context.
In `CodeMirror.multiplexingMode()` we define a new mode and pass the original one as the base, but the original mime type gets lost (sql-hint collects SQL keywords based on `doc.modeOption`).
I guess this modeOption isn't even something that should be called in `getKeywords()`, instead `cm.getModeAt()` would be appropriate. But even then we only know the correct mode but we don't know the mime type, which is necessary to know the SQL dialect.

I see two problems: One being the `getKeywords()` logic in sql-hint.js and the other that the multiplexer should pass through the base mime type, since currently any addon depending on a mime type won't work out of the box with the multiplexer.
Maybe I am just missing something, but it looks like the mime type is not stored anywhere else than in the modeOption and is not positional aware. Don't know if it'd be the right place to get it from the `cm.getModeAt()` result.

So currently, I don't have any idea for a solution. The only possible workaround is to override `doc.modeOption` after instantiating the editor with the multiplexing mode.

**Comment by Irongade** [comment]
Hello @neon-dev This work around is not clear to me, could you shine more light on this?

**Comment by Irongade** [comment]
Hi @marijnh I've been dealing with this issue for over a month now, I would appreciate if there's any known way to multiplex sql with javascript. 

If the code below is used, javascript fails to be highlighted between the `{{ }}` For example:

<img width="1125" alt="Screenshot 2022-10-12 at 13 29 54" src="https://user-images.githubusercontent.com/37867493/201497143-93aa49f6-fc13-49a0-ae6c-d7c5bf180135.png">


`
CodeMirror.defineMode("sql-js", function(config) {
  return CodeMirror.multiplexingMode(
    CodeMirror.getMode(config, "sql"),
    {
      open: "{{",
      close: "}}",
      mode: CodeMirror.getMode(config, {
        name: "javascript",
      }),
    },
  );
});
`

**Comment by marijnh** [comment]
The code you provide works fine for me. Possibly you're not loading the required scripts (javascript and sql mode, addon/mode/multiplex.js).

In any case, @-mentioning all maintainers is obnoxious. Don't do that.

**Comment by Irongade** [comment]
Once again, I would like to apologize for obnoxiously @-mentioning the maintainers, it won't happen again.

with regards to the issue, I wasn't saying there was a problem with the multiplexing system between SQL and javascript. the issue I was referring to was how SQL hinting fails to work with a multiplexed SQL-JS mode, which was why this issue was created in the first place. 

For example:
The code below represents adding SQL hints to codemirror using the SQL hint addon found in this file path `codemirror/addon/hint/sql-hint` which is called via `Codemirror.hint.sql`

```
import "codemirror/addon/hint/sql-hint";
import "codemirror/addon/hint/show-hint";
import "codemirror/addon/hint/show-hint.css";

showHint: (editor: CodeMirror.Editor): boolean => {
      editor.showHint({
        hint: CodeMirror.hint.sql,
        completeSingle: false,
        extraKeys: {
          [KeyboardShortcuts.CodeEditor.OpenAutocomplete]: "autocomplete",
        },
      });

      editor.setOption("hintOptions", {
        tables, // custom list of sql tables and column names
      });
      return true;
    },

```

This would normally work for a simple SQL mode with codemirror.

However an issue arises when we try to use this with a multiplexed SQL-JS codemirror mode i.e. The hinting of SQL based commands stop working e.g Where, From etc. It is important to note however that hinting of the tables and column names still work as expected.

Example of the multiplexed mode under consideration was given as an example above.

In summary, `Codemirror.hint.sql` only works for SQL mode but not multiplexed SQL-JS mode.

I would appreciate any workaround or insights into how this can be fixed. Many thanks. @marijnh

**Comment by marijnh** [comment]
>  ... javascript fails to be highlighted between the `{{` `}}`

Hm, I wonder where I got the idea that this was about highlighting.

But indeed, sql-hint makes some bad assumptions about how modes work. Attached patch should help.

**Comment by Irongade** [comment]
haha, I appreciate the sarcasm. Thank you for this patch @marijnh. This is highly appreciated. 😅

**Comment by neon-dev** [comment]
> I see two problems: One being the getKeywords() logic in sql-hint.js and the other that the multiplexer should pass through the base mime type, since currently any addon depending on a mime type won't work out of the box with the multiplexer.

First one is fixed by the patch, but isn't the SQL dialect still hardcoded? If so, this issue is not completely fixed.  
Sorry if I'm missing something, I haven't touched CM 5 for years.

**Comment by Irongade** [comment]
@neon-dev It seems you are correct. 

@marijnh I tried the patch but it seems the issue has not been resolved, Autocompletion of the SQL commands is still not being done. could this be a mime type issue as @neon-dev suggested?

**Comment by Irongade** [comment]
Hello Marijn, Apologies for the intrusion @marijnh, I wanted to make a plea and drop a gentle reminder to help look into this issue once more. I have tested it myself as well and I can confirm that the issue is not totally resolved. As @neon-dev suggested, it is possible that another step is needed here to completely resolve this issue.

I would sincerely appreciate any workaround or insights into how this can be fixed. Once again, apologies for the intrusion and Many Thanks. 🙏
